cmd/controller/main.go

```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"log"
+	"time"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+
+	// Uncomment the following line to load the gcp plugin (only required to authenticate against GKE clusters).
+	// _ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+
+	"github.com/knative/build-pipeline/pkg/controller"
+	"github.com/knative/build-pipeline/pkg/controller/task"
+	"github.com/knative/build-pipeline/pkg/controller/taskrun"
+	onclusterbuilder "github.com/knative/build/pkg/builder/cluster"
+	buildctrl "github.com/knative/build/pkg/controller/build"
+
+	buildclientset "github.com/knative/build-pipeline/pkg/client/clientset/versioned"
+	informers "github.com/knative/build-pipeline/pkg/client/informers/externalversions"
+	knativebuildclientset "github.com/knative/build/pkg/client/clientset/versioned"
+	knativeinformers "github.com/knative/build/pkg/client/informers/externalversions"
+	cachingclientset "github.com/knative/caching/pkg/client/clientset/versioned"
+	cachinginformers "github.com/knative/caching/pkg/client/informers/externalversions"
+	"github.com/knative/pkg/configmap"
+	"github.com/knative/pkg/logging"
+	"github.com/knative/pkg/logging/logkey"
+	"github.com/knative/pkg/signals"
+)
+
+const (
+	threadsPerController = 2
+	logLevelKey          = "controller"
+)
+
+var (
+	kubeconfig = flag.String("kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
+	masterURL  = flag.String("master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+)
+
+func main() {
+	flag.Parse()
+	loggingConfigMap, err := configmap.Load("/etc/config-logging")
+	if err != nil {
+		log.Fatalf("Error loading logging configuration: %v", err)
+	}
+	loggingConfig, err := logging.NewConfigFromMap(loggingConfigMap)
+	if err != nil {
+		log.Fatalf("Error parsing logging configuration: %v", err)
+	}
+	logger, _ := logging.NewLoggerFromConfig(loggingConfig, logLevelKey)
+	defer logger.Sync()
+	logger = logger.With(zap.String(logkey.ControllerType, logLevelKey))
+
+	logger.Info("Starting the Build Controller")
+
```

Comment:
For a *super* skeletal example to start from using the `knative/pkg` goodies check out https://github.com/mattmoor/warm-image.

This doesn't leverage things like the webhook (which Build hasn't yet adopted), but it shows the `Reconciler` bits.

Reply:
There are a skeleton implementation in [build](https://github.com/knative/build/tree/master/pkg/reconciler) but its not fully completed yet. `Serving` is a better example

Reply:
I'm going to leave this as is for now and refactor to using the new controller interfaces/libs in another PR

Reply:
isn't the point of this PR to refactor the controllers interfaces libs to the new style?

Reply:
> isn't the point of this PR to refactor the controllers interfaces libs to the new style?

This particular PR is about removing kubebuilder, but I agree that we want the controller code to settle in the style we want to continue with. Since this PR is so huge, I think we should go ahead with it as-is, then have a subsequent PR to update to the new style.

Hopefully I'm not underestimating the differences between this and the new style, but from what I can tell it seems like folks can go ahead and get unblocked if we merge this, then it's just a few changes to update to the latest.

(As part of updating to the latest I'd also like to write some docs about it b/c I haven't been able to find any.)

Reply:
ok, I agree this PR is getting to big to follow, we can fix up whatever needs changes after.